Add MemArena allocator

xf6wang opened the pull request:
`MemArena` is a memory allocator for process globals and transient arenas. Spans of memory within the arena can be allocated to amortize the cost of allocations. The idea is to alloc chunks of memory within the arena, freeze the allocations, copy the contents into a larger arena block, and thaw away the old memory contents. This effectively allows for coalescence into a single memory block. 

- `MemArena` is self expanding, meaning that it will always return a large enough `MemSpan`. 
- It maintains the underlying memory using `MemArenaBlock`s which mark generations of the arena and increase in size by factors of 2 unless explicitly specified.  
- Internal memory is allocated as a `Block` header following immediately with the underlying memory that the header manages.

SolidWallOfCode commented:
Add documentation in `internal-libraries`.  [This](https://solidwallofcode.github.io/MemArena.en.html) might be a good start.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,160 @@
+/** @file
+
```
Just a small comment for the file here?

SolidWallOfCode commented:
As an implementation, I had originally planned for this to be integrated in to the `IOBuffer` system but given the move toward removing the proxy allocators, its seems reasonable to just go straight to `malloc`. I've looked at using the `IOBuffer` structures, but I think those are not quite suitable for this use. While the linked list feature of `IOBufferBlock` would be useful, that would require 3 allocations per block (`IOBufferBlock`, `IOBufferData`, and the actual memory) which seems a bit overly complex.

xf6wang commented:
Updated with documentation for `MemArena`

SolidWallOfCode commented on lib/ts/Makefile.am:
```diff
@@ -271,7 +273,8 @@ test_tslib_SOURCES = \
 	unit-tests/test_string_view.cc \
 	unit-tests/test_TextView.cc \
-	unit-tests/test_MemSpan.cc
+	unit-tests/test_MemSpan.cc \
```
Do me a favor and move both `MemSpan` and `MemArena` so the test source files are in alphabetically order.

xf6wang replied:
done.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,160 @@
+    limitations under the License.
+ */
+
+#include "MemArena.h"
```
This should be `<ts/MemArena.h>` because that's how every thing else will include it.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,160 @@
+inline void
+MemArena::MemArenaBlock::free()
+{
+  if (data) {
```
`free` works fine on a `null` pointer. Don't bother to check.

xf6wang replied:
ok. done.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,160 @@
+  }
+  size      = 0;
+  allocated = 0;
+  next      = nullptr;
```
Why set this to `nullptr` but not `data`? Both or neither.

xf6wang replied:
`nullptr` for both.

SolidWallOfCode commented:
Also, add another method
```
/// @return @c true if @a ptr is in memory owned by this arena, @c false if not.
bool MemArena::contains(void* ptr) const;
```

xf6wang commented:
Updated with `MemArena::contains` which searches through internal blocks for ptr address.

SolidWallOfCode commented:
[approve ci]

SolidWallOfCode commented:
[approve ci clang-analyzer]

SolidWallOfCode commented:
[approve ci]

SolidWallOfCode commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -0,0 +1,145 @@
+Internals
++++++++++
+
+|MemArena| opperates in *generations*. Each generation marks a seperate internal block of memory. From a generation, memory can be allocated in the form of a :class:`MemSpan`. Once a generation is exhausted all avaliable memory, or is explicitly "frozen" via :func:`MemArena::freeze`, all allocations are frozen and a new generation must be created. Allocations made in previous generations can stil ...
```
This isn't correct. A generation can have any number of memory blocks. Generations are only created by explicit API calls. Every time a new block is needed it is twice the size of the previous block in the generation.

xf6wang replied:
Yea. I misunderstood what `MemArena` was supposed to do. I thought it was the case that each block was a new generation. It should be that you can allocate new internal blocks, but are in the same generation. Generations should only be marked with explicit calls to `freeze`. I updated the PR with new allocation logic.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,134 @@
+public:
+  /** Simple internal arena block of memory. Maintains the underlying memory.
+   */
+  struct MemArenaBlock {
```
Why not just `Block`? It's already syntactically scoped to `MemArena`.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,134 @@
+    void *data;
+    size_t size;
+    size_t allocated;
+    MemArenaBlock *next;
```
Perhaps use `std::shared_ptr` here and one in `MemArena`. Then discarding the string of blocks is a single assignment.

xf6wang replied:
Updated to use `shared_ptr`

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,134 @@
+  };
+
+  MemArena();
+  MemArena(size_t n);
```
Mark this `explicit`.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,134 @@
+      will be in a new generation.
+
+      @param n Number of bytes for new generation.
+        if @a n == 0. double in size.
```
No. If not specified the new block should be large enough to hold all existing allocations.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,134 @@
+private:
+  void create_new_generation();
+
+  const unsigned int DEFAULT_BLOCK_SIZE = 32000; ///< 32kb
```
`static constexpr size_t DEFAULT_BLOCK_SIZE = 1<<15;`. That might be a bit large to start.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,179 @@
+
+  MemArenaBlock *tmp = allocations;
+  while (tmp) {
+    if (ptr >= static_cast<char *>(tmp->data) && ptr < static_cast<char *>(tmp->data) + tmp->size) {
```
The `Block` class should have a `data()` method that returns a `char*` to its allocated data.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,179 @@
+MemArena::freeze(size_t n)
+{
+  // Get the mutex here to stop other allocations in internal memory blocks.
+  std::lock_guard<std::mutex> lock(arena_mutex);
```
What exactly is this locking against? Other calls to `freeze`?

xf6wang replied:
Only one thread should be able to alloc in the arena at a time so there's a lock. If there's a call to `freeze`, it should stop all other allocations before starting a new generation so it would need the lock here.

SolidWallOfCode replied:
I think we'll be non-thread safe. If the owner needs thread safety, it should provide it. In all of my envisioned use cases, this is how it is expected to work.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,179 @@
+    next_generation = false;
+  }
+
+  // keep creating generations memory blocks until we have something large enough
```
What happens to all those intermediate blocks? Is it really efficient to allocate these until one is big enough?

xf6wang replied:
Default is to try double in size. If it's not big enough, instead of doubling in size, the next internal block will just be `n` bytes.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,179 @@
+{
+  MemArenaBlock *tmp = allocations;
+  MemArenaBlock *next;
+  while (tmp) {
```
This looks very similar to `thaw`. Why duplicate? As noted elsewhere, `std::shared_ptr` is an alternative that removes all this code.

xf6wang replied:
done.

xf6wang commented:
@SolidWallOfCode I updated the PR quite a bit. 
- I had previously misunderstood how the generations were supposed to work. Now, generations are only marked with calls to `freeze`.
-  It also uses `shared_ptr`s to manage the allocations. I checked on `valgrind` that there are no leaks.
- If you `freeze` then immediately `thaw`(no allocations), it empties the arena. This allows you to actually scale down because you can `freeze(smaller)` and `thaw` and allocate in a smaller arena. 
- Added a couple of helper functions (`empty`, `allocated_size`, `unallocated_size`)
- Updated docs :)

xf6wang commented:
Removed mutex in `MemArena`

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+{
+  size      = n;
+  allocated = 0;
+  next      = nullptr;
```
Shouldn't this be a constructor? Under what circumstances would you construct an instance of `Block` and not call `init`?
```
MemArena::Block::Block(size_t n) : size(n) {}
// next is constructed to nullptr because it's a shared_ptr, and allocated should be class initialized to 0.
```

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+    t |= t >> 8;
+    t |= t >> 16;
+    t++;
+    n = t - sizeof(Block) - 16;
```
Make `16` a class constant, e.g.
```
static constexpr size_t ALLOC_HEADER_SIZE = 16;
```

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+    t |= t >> 8;
+    t |= t >> 16;
+    t++;
+    n = t - sizeof(Block) - 16;
```
Also, what is the point of computing `n` here? It doesn't seem to be used.

xf6wang replied:
`n` is the actual size available that the `Block` can allocate out. It is used to initialize the `Block` as well as set up `generation_size` and `arena_size`.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  auto block = std::shared_ptr<Block>(new (tmp) Block); // placement new
+  block->init(n);
```
As noted elsewhere, this should be blended in to the constructor.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+    t |= t >> 2;
+    t |= t >> 4;
+    t |= t >> 8;
+    t |= t >> 16;
```
It seems to me that adding `t |= t >> 32;` would work to cover the 64 bit case. You can see in the pattern that it's doubling the number of leading 1 bits each time, so one more step is enough.

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+
+MemArena::~MemArena()
+{
+  empty();
```
I think this is unnecessary. The `shared_ptr` instances will clear themselves on destruct, which should trigger cleaning up all the memory. Clearing the other member is pointless in the destructor.

xf6wang replied:
Removed.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,159 @@
+    char *data();
+  };
+
+  explicit MemArena();
```
No `explicit` for default constructors.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,159 @@
+  std::shared_ptr<Block> allocations = nullptr; ///< Management of memory -- entire list
+  std::shared_ptr<Block> generation  = nullptr; ///< Marks current generation
+
+  std::shared_ptr<Block> current = nullptr; ///< Allocate memory from this
```
How is this different from `generation`?

SolidWallOfCode replied:
I would tend to have `current` for the current generation and `previous` or `frozen` for the previous generation(s).

xf6wang replied:
Maybe this is poorly named. `current` is supposed to point to the most recent block within a generation. Under normal circumstances, this is the `Block` to allocate memory from. `generation` points to the beginning of the current generation. The reason I have them separated is:
                       `gen`                               `cur`      
( block ) -> ( block ) -> ( block ) -> ( block )

I just set `allocations` to `generation` to deallocate frozen generations.

SolidWallOfCode commented on lib/ts/MemArena.h:
```diff
@@ -0,0 +1,159 @@
+  size_t total_alloc     = 0;                  ///< Total number of bytes allocated in the arena -- allocated
+  size_t prev_alloc      = 0;                  ///< Total allocations before current generation -- allocated
+
+  bool next_generation   = false;
```
Why have this, instead of just checking if `generation` is `nullptr`?

xf6wang replied:
good point. changed.

xf6wang commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+inline char *
+MemArena::Block::data()
+{
+  return reinterpret_cast<char *>(this + sizeof(Block));
```
Fixed.

xf6wang commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,232 @@
+    tmp = ats_malloc(t - 16);
+  }
+
+  auto block = std::shared_ptr<Block>(new (tmp) Block); // placement new
```
No good reason. Changed.

xf6wang commented:
Updated docs and unit tests

SolidWallOfCode commented on lib/ts/MemArena.cc:
```diff
@@ -0,0 +1,221 @@
+    generation_size += n;
+
+    if (back) { // append
+      current->next = block;
```
Here, if you did
```
block->next = current;
current = block
```
The chain would be anchored by the most recent block and you would only need that to both find the most recent block and maintain the chain. Then `current` and `allocations` would be the same thing. For a "custom" block for an  allocation, you do
```
if (current) {
  block->next = current->next;
  current->next = block;
} else {
  current = block;
}
```
This leaves `current` the current block. Then you only need `current` and `generation`.

xf6wang replied:
Good point. Removed `allocations`. List now grows from head. Thawing just nulls `generation->next` if `generation` exists. Updated docs.
